#**Maestría en Inteligencia Artificial Aplicada (MNA)**
##**Procesamiento de Lenguaje Natural (NLP)**
###Prof. Luis Eduardo Falcón Morales
###**Semana 4 - Ejercicios Complementarios**
### Matriz Documento-Término y Tf-idf.

###**NOTA: Estos ejercicios son simplemente para que repases los conceptos de esta semana.**
###**No es una tarea o actividad que debas entregar.**

Comentamos que los diccionarios son una parte importante que permiten asociar a cada token su freuencia de aparición en los documentos de texto.

Sin embargo, esto no es suficiente para medir la importancia y sobre todo, para asociar algún significado al token con base a los demás tokens que se utilizan en una frase o documento.

Para ello, existen la matriz llamada documento-término y la matriz tf-idf.

La matriz documento-término (abreviada DTM pos sus siglas en inglés, Documen-Term-Matrix) es una matriz donde cada renglón representa un documento o enunciado y cada columna son los términos o tokens.

A partir de dicha matriz se puede obtener a su vez la llamada matriz término-frecuencia y frecuencia de documento inversa. En general se le llama por sus siglas en inglés tf-idf, term-frequency-inverse-document-frequency. 

La matriz tf-idf trata de obtener una representación de los tokens y los documentos, donde pondera la aparición de los tokens como importantes cuando aparece con frecuencia; pero no demasiado, porque si aparece demasiado castiga a dicho token disminuyendo su peso. Es decir, es algo análogo al concepto de los stopwords, donde un token muestra su importancia cuando aparece frecuentemente, pero no demasiado frecuente porque entonces se hace trivial su aparición.

Veamos estos conceptos con un ejercicio.

Retomemos el ejercicio de la semana pasada.

Recordemos que estamos suponiendo que tenemos la siguiente lista de comentarios en un problema de clasificación, donde cada uno está etiquetado como comentario positivo con un 1, o como comentario negativo con un 0.

In [1]:
import re

In [2]:
docs = [['Muy bien hecho lo hecho.', 1],
        ['¡Excelente trabajo muchacho!', 1],
        ['Sigue como hasta ahora muchacho.', 1],
        ['Lo hiciste muy bien.', 1],
        ['¡¡Excelente resultado!!', 1],
        ['¡¡¡Pésimo trabajo muchacho!!!', 0],
        ['Que mal desempeño.', 0],
        ['Muy mal hecho muchacho.', 0],
        ['¡Que trabajo tan pobre!', 0],
        ['No está bien, nada bien, muchacho.', 0]]

Apliquemos el proceso de limpieza y tokenización usual:

In [3]:
Xclean = []  
mis_stopwords = ['a', 'lo', 'que', 'está', 'hasta', 'el', 'la']

for doc in docs:
  words = re.sub(r'[^a-zA-ZáéíóúüñÁÉÍÓÚÜÑ]', ' ', doc[0])   
  words = re.sub(r'\s{2,}', ' ', words.strip())      
  words = words.lower().split()
  words = [ w for w in words if w not in mis_stopwords]
  Xclean.append(words)


Xx = []  
for x in Xclean:
  Xx.append(' '.join(x))


y = [t for _,t in docs]



# Xclean : lista de listas: documentos tokenizados y procesados.
# Xx : lista de strings: cada documento es un string.
# y : etiquetas de cada documento.

In [4]:
print(Xclean)
print(Xx)
print(y)

[['muy', 'bien', 'hecho', 'hecho'], ['excelente', 'trabajo', 'muchacho'], ['sigue', 'como', 'ahora', 'muchacho'], ['hiciste', 'muy', 'bien'], ['excelente', 'resultado'], ['pésimo', 'trabajo', 'muchacho'], ['mal', 'desempeño'], ['muy', 'mal', 'hecho', 'muchacho'], ['trabajo', 'tan', 'pobre'], ['no', 'bien', 'nada', 'bien', 'muchacho']]
['muy bien hecho hecho', 'excelente trabajo muchacho', 'sigue como ahora muchacho', 'hiciste muy bien', 'excelente resultado', 'pésimo trabajo muchacho', 'mal desempeño', 'muy mal hecho muchacho', 'trabajo tan pobre', 'no bien nada bien muchacho']
[1, 1, 1, 1, 1, 0, 0, 0, 0, 0]


In [5]:
import pandas as pd

In [6]:
# Generamos nuestro vocabulario:

vocab = []
for x in Xclean:     
  vocab.extend(x)    
  vocab = list(set(vocab))

vocab.sort()  

print(len(vocab))
print(vocab)

18
['ahora', 'bien', 'como', 'desempeño', 'excelente', 'hecho', 'hiciste', 'mal', 'muchacho', 'muy', 'nada', 'no', 'pobre', 'pésimo', 'resultado', 'sigue', 'tan', 'trabajo']


Así, con 18 tokens se deberá procesar y analizar los documentos.

##**Matriz Documento-Término: Document-Term-Matrix : DTM**

###Comentamos que existe una gran variedad de librerías para obtener las diferentes representaciones matriciales de los documentos de texto.

###En particular, haciendo uso de CountVectorizer() de sklearn, podemos obtener la matriz DTM. 

###Si generaste tu vocabulario puedes usarlo para generar las matrices, de lo contrario puedes dejar que el propio método lo genere. Sin embargo, conviene trabajarlo uno personalmente, sobre todo por lo realizado con regex.

Veamos este caso sencillo:

In [7]:
from sklearn.feature_extraction.text import CountVectorizer

In [8]:
vectorizer = CountVectorizer(vocabulary= vocab)  

DTM = vectorizer.fit_transform(Xx)

print('matriz DTM:')
print(DTM.toarray())

matriz DTM:
[[0 1 0 0 0 2 0 0 0 1 0 0 0 0 0 0 0 0]
 [0 0 0 0 1 0 0 0 1 0 0 0 0 0 0 0 0 1]
 [1 0 1 0 0 0 0 0 1 0 0 0 0 0 0 1 0 0]
 [0 1 0 0 0 0 1 0 0 1 0 0 0 0 0 0 0 0]
 [0 0 0 0 1 0 0 0 0 0 0 0 0 0 1 0 0 0]
 [0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0 0 1]
 [0 0 0 1 0 0 0 1 0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 1 0 1 1 1 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 1]
 [0 2 0 0 0 0 0 0 1 0 1 1 0 0 0 0 0 0]]


###Podemos asociar el nombre de cada columna/token del vocabulario para mayor claridad, por ejemplo, nuevamente para el primer comentario:

In [9]:
count_tokens = vectorizer.get_feature_names_out()
df_countvect = pd.DataFrame(data = DTM.toarray(),  columns = count_tokens)  
print(df_countvect.iloc[0:1,0:]) 

   ahora  bien  como  desempeño  excelente  hecho  hiciste  mal  muchacho  \
0      0     1     0          0          0      2        0    0         0   

   muy  nada  no  pobre  pésimo  resultado  sigue  tan  trabajo  
0    1     0   0      0       0          0      0    0        0  


###La dimensión de esta document-term-matrix, DTM, es $(renglones\times columnas)=(Total De Comentarios \times Total De Tokens)$. Es decir, $10\times 18$ en nuestro ejemplo.

###Por ejemplo, el primer comentario es 'muy bien hecho hecho' y los índices de los tokens 'muy', 'bien' y 'hecho' son 9, 1 y 5. Y como 'hecho' aparece 2 veces en el primer comentario su vector asociado es como se muestra en el primer renglón de la matriz DTM.

###Por otro lado, obervamos que la matriz DTM tiene una gran cantidad de ceros. En general así sucede, porque cada comentario hace uso de una cantidad mínima de palabras del vocabulario. Este tipo de matrices son llamadas dispersas (sparse matrix).

###De manera predeterminada CountVectorizer() nos regresa una matriz dispersa. En este ejemplo hicimos uso del método "toarray()" para mostrar la matriz en su formato estándar. Sin embargo, esto no debes hacerlo en general, ya que son matrices muy muy grandes y tratar de visualizarlas de la manera estándar puede hacer uso de toda la memoria RAM.

###En particular, el formato predeterminado del tipo de matriz dispersa es la CSR (Compressed-Sparse-Row, por sus siglas en inglés).

###Lo podemos verificar:

In [10]:
DTM

<10x18 sparse matrix of type '<class 'numpy.int64'>'
	with 31 stored elements in Compressed Sparse Row format>

Tenemos 31 elementos no cero, de un total de 180 entradas de la matriz.

Es decir:

In [11]:
DTM.count_nonzero() / (DTM.shape[0] * DTM.shape[1])

0.17222222222222222

Es decir, en este pequeño ejemplo el 17% de los registros son no cero y el 83% son entradas de valor cero. Así, diremos que la matriz DTM tiene una dsipersidad (sparsity en inglés) del 83%. 

En la práctica, es usual que la dispersidad de las matrices DTM esté muy por arriba del 95%. Por ello la conveniencia de utilizar este tipo de formatos dispersos.

##**Matriz Tf-idf**

Como comentamos en las diapositivas de la teoría de esta semana, la matriz Tf-idf pondera cada token por frecuencia de aparición, pero si aparece "demasiadas" veces, se le penaliza disminuyendo su peso.

In [12]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidfvectorizer = TfidfVectorizer(vocabulary=vocab, norm=None)
DTM_train_tfidf = tfidfvectorizer.fit_transform(Xx)   

In [13]:
tfidf_tokens = tfidfvectorizer.get_feature_names_out()
df_tfidfvect = pd.DataFrame(data = DTM_train_tfidf.toarray(),  columns = tfidf_tokens)  
print(df_tfidfvect.iloc[0:1,0:]) 

   ahora      bien  como  desempeño  excelente     hecho  hiciste  mal  \
0    0.0  2.011601   0.0        0.0        0.0  4.598566      0.0  0.0   

   muchacho       muy  nada   no  pobre  pésimo  resultado  sigue  tan  \
0       0.0  2.011601   0.0  0.0    0.0     0.0        0.0    0.0  0.0   

   trabajo  
0      0.0  


Para este primer comentario vemos que la palabra "hecho" es la considerada que da mayor información con el mayor valor. Y su importancia en este comentario se justifica por aparecer dos veces en este. A los tokens "bien" y "muy" los poderó de manera igual, que en este caso ambos tokens aparecen una sola vez.

En la preentación de este semana definimos de manera detallada las fórmulas para calcular el valor de $tf\times idf$.

Sin embargo, existen muchas variantes en la manera en que se definen estos términos, en particular la $idf$. En general, las variaciones llegan a ser diversas, pero todas nos llevan en general a resultados muy equivalentes cuando usamos esta información con modelos de aprendizaje de máquinas.

Se se revisa la documentación de TfidfVectorizer() vemos la manera en que calcula dichos valores.

Puedes revisar la documentación en las siguientes dos ligas:

https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfTransformer.html#sklearn.feature_extraction.text.TfidfTransformer

https://scikit-learn.org/stable/modules/feature_extraction.html#text-feature-extraction



En particular, la fórmula que utiliza TfidfVectorizer() con el argumento "norm=None" que usamos arriba, es como sigue. Calculando el valor para el primer comentario y el token "hecho".

$tf\times idf = tf\times ln(\frac{D+1}{n_t+1})=2\times \{ln(\frac{10+1}{2+1})+1\}=4.598566$.

Que es el resultado que se muestra en la salida que mostramos más arriba.


Lo importante es que estos valores de las matrices DTM y DTM_tfidf obtenidas con un conjunto de entrenamiento, pueden utilizarse para entrenar un modelo y después validarlo con los conjuntos de validación y de prueba.

A manera de ejemplo, supongamos que deseamos aplicar estos modelos con regresión logística.

In [14]:
from sklearn.linear_model import LogisticRegression

In [15]:
modeloCount = LogisticRegression(max_iter=100, C=1.)
modeloCount.fit(DTM, y)

LogisticRegression()

In [16]:
print('\nModelo Counter con regresión Logística: Train-accuracy: %.2f%%' % (100*modeloCount.score(DTM, y)))


Modelo Counter con regresión Logística: Train-accuracy: 100.00%


Y de manera análoga con la matriz Tf-idf:

In [17]:
modeloTfidf = LogisticRegression(max_iter=100, C=1.)
modeloTfidf.fit(DTM_train_tfidf, y)
print('\nModelo Tfidf con Regresión Logística: Train-accuracy: %.2f%%' % (100*modeloTfidf.score(DTM_train_tfidf, y)))


Modelo Tfidf con Regresión Logística: Train-accuracy: 100.00%


Sobra decir que este modelo esta sobre-entrenado porque solo estamos usando un mismo conjunto para entrenar y validar, pero es solo a manera de ejemplo. Cuando se tenga el conjunto de validación y de prueba deberás hacer la validación de manera adecuada.

##**Fin de la Parte 2 de los ejercicios complementarios de la Semana 3**